### PR TITLE
[BugFix] add dropped flag to avoid db dropped after db got from catalog

### LINKED ISSUE #10346: [BUG] FE Start failed when replaying insert overwrite jobs
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'

### Real behavior (Required)
```java
2022-08-23 15:48:47,693 ERROR (stateChangeExecutor|73) [EditLog.loadJournal():884] Operation Type 10002
java.lang.NullPointerException: null
        at com.starrocks.server.LocalMetastore.replayAddPartition(LocalMetastore.java:1351) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayAddPartition(GlobalStateMgr.java:1794) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.EditLog.loadJournal(EditLog.java:206) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:1670) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:1628) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.transferToLeader(GlobalStateMgr.java:936) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.access$100(GlobalStateMgr.java:266) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$1.transferToLeader(GlobalStateMgr.java:591) [starrocks-fe.jar:?]
        at com.starrocks.ha.StateChangeExecutor.runOneCycle(StateChangeExecutor.java:86) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
```

### StarRocks version (Required)
 - main

### LINKED ISSUE #10381: [BUG] FE Start failed when replaying insert overwrite state change
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'

### Real behavior (Required)
```java
2022-08-24 10:33:12,346 ERROR (stateChangeExecutor|73) [EditLog.loadJournal():884] Operation Type 10096
java.lang.NullPointerException: null
        at com.starrocks.load.InsertOverwriteJobRunner.<init>(InsertOverwriteJobRunner.java:66) ~[starrocks-fe.jar:?]
        at com.starrocks.load.InsertOverwriteJobManager.replayInsertOverwriteStateChange(InsertOverwriteJobManager.java:151) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.EditLog.loadJournal(EditLog.java:857) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:1670) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:1628) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.transferToLeader(GlobalStateMgr.java:936) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.access$100(GlobalStateMgr.java:266) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$1.transferToLeader(GlobalStateMgr.java:591) [starrocks-fe.jar:?]
        at com.starrocks.ha.StateChangeExecutor.runOneCycle(StateChangeExecutor.java:86) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
```

### StarRocks version (Required)
 - main

### PR DESCRIPTION
**Pull request by gengjun-git**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #10381
Fixes #10346

## Problem Summary(Required) ：

Many places call db.writelock after getting db from catalog, and then write bdb log. But after getting db from catalog, the db may still be dropped, which will cause the NPE error when replay bdb log. So add a dropped flag to cover this case. The set and check of dropped flag should be in the db lock.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 27 / 63 (42.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/HiveTable.java|0|2|00.00%|[287, 288]|
|:large_blue_circle:|com/starrocks/leader/ReportHandler.java|0|1|00.00%|[937]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|0|3|00.00%|[203, 204, 205]|
|:large_blue_circle:|com/starrocks/consistency/CheckConsistencyJob.java|0|3|00.00%|[256, 257, 258]|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|0|2|00.00%|[723, 724]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|6|16|37.50%|[537, 538, 566, 1269, 4176, 4391, 4392, 4432, 4528, 4529]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|6|12|50.00%|[136, 172, 240, 419, 518, 629]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|2|4|50.00%|[156, 549]|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|1|2|50.00%|[582]|
|:large_blue_circle:|com/starrocks/load/PartitionUtils.java|1|2|50.00%|[33]|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2.java|1|2|50.00%|[220]|
|:large_blue_circle:|com/starrocks/load/loadv2/BrokerLoadJob.java|1|2|50.00%|[287]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|1|2|50.00%|[346]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|1|2|50.00%|[1206]|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|1|2|50.00%|[911]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/AlterReplicaTask.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/Database.java [hunk 1]
```diff
@@ -99,6 +99,8 @@ public class Database extends MetaObject implements Writable {
 
     private long lastSlowLockLogTime = 0;
 
+    private boolean dropped = false;
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Database.java:102]: should be atomic variable?

**Review by ABingHuang — DISMISSED** [review]
LGTM

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 30 / 61 (49.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/HiveTable.java|0|2|00.00%|[287, 288]|
|:large_blue_circle:|com/starrocks/consistency/CheckConsistencyJob.java|0|3|00.00%|[256, 257, 258]|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|0|2|00.00%|[723, 724]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|6|16|37.50%|[537, 538, 566, 1269, 4176, 4391, 4392, 4432, 4528, 4529]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|6|12|50.00%|[136, 172, 240, 419, 518, 629]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|2|4|50.00%|[156, 549]|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|1|2|50.00%|[582]|
|:large_blue_circle:|com/starrocks/load/loadv2/BrokerLoadJob.java|1|2|50.00%|[287]|
|:large_blue_circle:|com/starrocks/task/AlterReplicaTask.java|1|2|50.00%|[169]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|1|2|50.00%|[346]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|1|2|50.00%|[1206]|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|1|2|50.00%|[911]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/PartitionUtils.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/leader/ReportHandler.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 27 / 61 (44.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/HiveTable.java|0|2|00.00%|[287, 288]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|0|3|00.00%|[203, 204, 205]|
|:large_blue_circle:|com/starrocks/consistency/CheckConsistencyJob.java|0|3|00.00%|[256, 257, 258]|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|0|2|00.00%|[723, 724]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|6|16|37.50%|[537, 538, 566, 1269, 4176, 4391, 4392, 4432, 4528, 4529]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|6|12|50.00%|[136, 172, 240, 419, 518, 629]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|2|4|50.00%|[156, 549]|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|1|2|50.00%|[582]|
|:large_blue_circle:|com/starrocks/load/PartitionUtils.java|1|2|50.00%|[33]|
|:large_blue_circle:|com/starrocks/load/loadv2/BrokerLoadJob.java|1|2|50.00%|[287]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|1|2|50.00%|[346]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|1|2|50.00%|[1206]|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|1|2|50.00%|[911]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/leader/ReportHandler.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/task/AlterReplicaTask.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/task/AlterReplicaTask.java [hunk 2]
```diff
@@ -164,6 +164,10 @@ public void handleFinishAlterTask() throws MetaNotFoundException {
 
         db.writeLock();
```
Comment (padmejin) [inline fe/fe-core/src/main/java/com/starrocks/task/AlterReplicaTask.java:165]: How about validate if database is dropped when calling db.xxLock()?

Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/task/AlterReplicaTask.java]: done

File: fe/fe-core/src/main/java/com/starrocks/alter/Alter.java [hunk 3]
```diff
@@ -129,7 +129,9 @@ public void processCreateMaterializedView(CreateMaterializedViewStmt stmt)
         // check db quota
         db.checkQuota();
 
-        db.writeLock();
+        if (!db.writeLockAndExist()) {
```
Comment (padmejin) [inline fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:132]: My initial suggestion is to change the behavior of `db.writeLock()` and throw an exception when calling `db.writeLock()` on a dropped database. All the callers can handle this exception as it wishes.

**Comment by wanpengfei-git** [comment]
run all

**Pull request closed without merging**